### ISSUE TITLE
duplicate key value violates unique constraint "events_pkey" error while creating multiple workflows

### ISSUE THREAD

**Issue by invidian** [tinkerbell/tink#379]
## Expected Behaviour
Workflows can be created in parallel without errors.

## Current Behaviour
Sometimes, when many workloads are created in parallel, following error is returned:
```
Error: creating workflow: rpc error: code = Unknown desc = failed to create workflow: INSERT in to workflow: pq: duplicate key value violates unique constraint "events_pkey"
```

Logs from server:
```
tink-server_1  | {"level":"error","ts":1607592790.282655,"caller":"grpc-server/workflow.go:159","msg":"UPDATE: pq: duplicate key value violates unique constraint \"events_pkey\"","service":"github.com/tinkerbell/tink","detail":"Key (created_at)=(2020-12-10 09:33:10.278354+00) already exists.","where":"SQL statement \"INSERT INTO events(resource_id, resource_type, event_type, data) VALUES (new.id, 'WORKFLOW', 'DELETED', row_to_json(new))\"\nPL/pgSQL function insert_workflow_event() line 9 at SQL statement","error":"UPDATE: pq: duplicate key value violates unique constraint \"events_pkey\"","errorVerbose":"pq: duplicate key value violates unique constraint \"events_pkey\"\nUPDATE\ngithub.com/tinkerbell/tink/db.TinkDB.DeleteWorkflow\n\t/usr/myapp/db/workflow.go:392\ngithub.com/tinkerbell/tink/grpc-server.(*server).DeleteWorkflow\n\t/usr/myapp/grpc-server/workflow.go:152\ngithub.com/tinkerbell/tink/protos/workflow._WorkflowService_DeleteWorkflow_Handler.func1\n\t/usr/myapp/protos/workflow/workflow.pb.go:1991\ngithub.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1\n\t/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107\ngithub.com/tinkerbell/tink/protos/workflow._WorkflowService_DeleteWorkflow_Handler\n\t/usr/myapp/protos/workflow/workflow.pb.go:1993\ngoogle.golang.org/grpc.(*Server).processUnaryRPC\n\t/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1194\ngoogle.golang.org/grpc.(*Server).handleStream\n\t/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1517\ngoogle.golang.org/grpc.(*Server).serveStreams.func1.2\n\t/go/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:859\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1374"}
```
## Possible Solution

Perhaps server should retry the write when this occurs. This is very similar to #338

## Steps to Reproduce (for bugs)

1. `git clone https://github.com/kinvolk/terraform-provider-tinkerbell && cd terraform-provider-tinkerbell && git checkout 71d569b946b6b93ecef12c39b29df997c233fd37 && git cherry-pick
a2672c384f32324da6b62c359445687fd09bd9c4`
2. `docker-compose -f test/docker-compose.yml up -d`
3. `TF_ACC=1 TINKERBELL_GRPC_AUTHORITY=127.0.0.1:42113 TINKERBELL_CERT_URL=http://127.0.0.1:42114/cert go test -v -run=^TestAccWorkflow_.*$ -count 100 ./tinkerbell/`

**Comment by thebsdbox** [comment]
This also looks related to events @gauravgahlot ?

**Comment by gauravgahlot** [comment]
> This also looks related to events @gauravgahlot ?

Not directly, but yes it does. The reason being, we do not insert into the events table from the server. All event entries are pushed from the triggers execute at the DB layer. That being said, this surely needs to be fixed at the root level.

**Comment by invidian** [comment]
>Not directly, but yes it does. The reason being, we do not insert into the events table from the server. All event entries are pushed from the triggers execute at the DB layer when you CUD a resource (hardware, template, workflow). That being said, this surely needs to be fixed at the root level.

This sounds like an architecture issue to me. This means event entries cannot be tested without a database for example.

**Comment by gianarb** [comment]
> This sounds like an architecture issue to me. This means event entries cannot be tested without a database for example.

Can you elaborate? You can mock almost everything

**Comment by invidian** [comment]
>Can you elaborate? You can mock almost everything

I didn't look much at the code, so please correct me if I'm wrong, but as far as I see, events are created here: https://github.com/tinkerbell/tink/blob/9c52cca562bcb70c89228db41ed25a0ef6ca5f52/db/migration/202010071530-init-events-table-and-triggers.go#L50

This means, without a database engine running, you cannot verify, that events are being created on certain operations.

In most of the cases, stored procedures are something which shouldn't be used, with some exceptions (like creating views to give unprivileged user access to sub-set of table). More about that for example here: https://stackoverflow.com/questions/1761601/is-the-usage-of-stored-procedures-a-bad-practice#:~:text=It%20isn't%20possible%20to,promote%20bad%20configuration%20management%20practices.&text=Stored%20procedures%20are%20inflexible.

**Comment by gianarb** [comment]
I think @mmlb has opinions about what should be used in the database, I am not an expert.

Anyway, your concern was about mocking and I am not sure I understand it, what can't you properly test? The event fire can be mocked from my understanding

**Comment by invidian** [comment]
>mmm mocking the db.Driver you can test it

Right. But then still your SQL procedure might be different than your mock.

But I think this is out of scope for this issue.

**Comment by invidian** [comment]
Ah, edited reproducing steps, as this only occurs when latest version of Tinkerbell is used, so please checkout tinkerbell/terraform-provider-tinkerbell@71d569b946b6b93ecef12c39b29df997c233fd37 while testing.

**Comment by gianarb** [comment]
Ok thanks! I changed the priority to P1 because we have to figure out what is going before we release a new version of Sandbox. @gauravgahlot can you have a look, please?

**Comment by gauravgahlot** [comment]
Yeah already. In conversation with @invidian for the same.
